WalletWasabi/Wallets/FilterProcessor/BlockDownloadService.cs

```diff
@@ -0,0 +1,370 @@
+				{
+					continue;
+				}
+
+				if (activeTasks.Count == MaximumParallelTasks)
+				{
+					Task task = await Task.WhenAny(activeTasks).ConfigureAwait(false);
+					activeTasks.Remove(task);
+				}
+			}
+		}
+		catch (OperationCanceledException)
+		{
+			Logger.LogDebug("Block-downloading service execution was stopped.");
+		}
+		catch (Exception ex)
+		{
+			// This shouldn't happen.
+			Logger.LogError(ex);
+			throw;
+		}
+		finally
+		{
+			// Mark everything as cancelled because the service is shutting down (either gracefully or forcibly).
+			lock (Lock)
+			{
+				while (BlocksToDownload.TryDequeue(out Request? request, out _))
+				{
+					request.Tcs.TrySetResult(CanceledResult.Instance);
+				}
+			}
+		}
+	}
+
+	private async Task HandleSingleBlockTaskAsync(Request request, CancellationToken cancellationToken)
+	{
+		RequestResponse response = await DownloadSingleBlockAsync(request, cancellationToken).ConfigureAwait(false);
+
+		if (response.Result is SuccessResult)
+		{
+			request.Tcs.TrySetResult(response.Result);
+		}
+		else
+		{
+			Logger.LogDebug($"Attempt to download block {request.BlockHash} (height: {request.Priority.BlockHeight}) failed.");
+
+			// The block might have been removed concurrently if a reorg occurred.
+			if (!request.Tcs.TrySetResult(response.Result))
+			{
+				IResult opResult = await request.Tcs.Task.ConfigureAwait(false);
+				Logger.LogDebug($"Failed to set result for block '{request.BlockHash}' (height: {request.Priority.BlockHeight}). Result is already: {opResult}");
+			}
+		}
+	}
+
+	/// <summary>
+	/// Downloads a single block from a block source, or gets the block from the file-system cache if available.
+	/// </summary>
+	private async Task<RequestResponse> DownloadSingleBlockAsync(Request request, CancellationToken cancellationToken)
+	{
+		Logger.LogTrace($"Trying to download {request.BlockHash} (height: {request.Priority.BlockHeight}).");
+
+		try
+		{
+			// Try get the block from the file-system storage.
+			Block? block = await FileSystemBlockRepository.TryGetAsync(request.BlockHash, cancellationToken).ConfigureAwait(false);
+			if (block is not null)
+			{
+				return new RequestResponse(request, new SuccessResult(block, EmptySourceData.FileSystemCache));
+			}
+
+			SuccessResult? successResult = null;
+			ISourceData? failureSourceData = null;
+
+			if (request.SourceRequest is FullNodeSourceRequest)
+			{
+				if (TrustedFullNodeBlockProviders.Length == 0)
+				{
+					return new RequestResponse(request, NoSuchProviderResult.Instance);
+				}
+
+				foreach (IBlockProvider blockProvider in TrustedFullNodeBlockProviders)
+				{
+					block = await blockProvider.TryGetBlockAsync(request.BlockHash, cancellationToken).ConfigureAwait(false);
+
+					if (block is not null)
+					{
+						successResult = new(block, EmptySourceData.TrustedFullNode);
+						break;
+					}
+				}
+
+				if (successResult is null)
+				{
+					failureSourceData = EmptySourceData.TrustedFullNode;
```

Comment:
There is no fallback to P2P is the trusted fails? We are not using this in the integration right now, correct? We use automatic.

Reply:
What happens in case of failure after this with the synchronization? What's the plan?

Reply:
ACK